src/txn/txn_nsnap.c

```diff
@@ -0,0 +1,372 @@
+	}
+
+	/*
+	 * The new ID will be none if we are removing all named snapshots
+	 * which is the default behavior of this loop.
+	 */
+	new_nsnap_oldest = WT_TXN_NONE;
+	if (name != NULL) {
+		STAILQ_FOREACH(last, &txn_global->nsnaph, q) {
+			if (WT_STRING_MATCH(last->name, name->str, name->len))
+				break;
+			prev = last;
+		}
+		if (last == NULL)
+			WT_RET_MSG(session, EINVAL,
+			    "Named snapshot '%.*s' for drop not found",
+			    (int)name->len, name->str);
+
+		if (!inclusive) {
+			/* We are done if a drop until points to the head */
+			if (prev == 0)
+				return (0);
+			last = prev;
+		}
+
+		if (STAILQ_NEXT(last, q) != NULL)
+			new_nsnap_oldest = STAILQ_NEXT(last, q)->snap_min;
+	}
+
+	do {
+		nsnap = STAILQ_FIRST(&txn_global->nsnaph);
+		WT_ASSERT(session, nsnap != NULL);
+		STAILQ_REMOVE_HEAD(&txn_global->nsnaph, q);
+		__nsnap_destroy(session, nsnap);
+	/* Last will be NULL in the all case so it will never match */
+	} while (nsnap != last && !STAILQ_EMPTY(&txn_global->nsnaph));
+
+	/* Now that the queue of named snapshots is updated, update the ID */
+	txn_global->nsnap_oldest_id = new_nsnap_oldest;
+
+	return (ret);
+}
+
+/*
+ * __wt_txn_named_snapshot_begin --
+ *	Begin an named in-memory snapshot.
+ */
+int
+__wt_txn_named_snapshot_begin(WT_SESSION_IMPL *session, const char *cfg[])
+{
+	WT_CONFIG_ITEM cval;
+	WT_DECL_RET;
+	WT_NAMED_SNAPSHOT *nsnap, *nsnap_new;
+	WT_TXN *txn;
+	WT_TXN_GLOBAL *txn_global;
+	const char *txn_cfg[] =
+	    { WT_CONFIG_BASE(session, WT_SESSION_begin_transaction),
+	      "isolation=snapshot", NULL };
+	int started_txn;
+
+	started_txn = 0;
+	nsnap_new = NULL;
+	txn_global = &S2C(session)->txn_global;
+	txn = &session->txn;
+
+	WT_RET(__wt_config_gets_def(session, cfg, "name", 0, &cval));
+	WT_ASSERT(session, cval.len != 0);
+
+	if (!F_ISSET(txn, WT_TXN_RUNNING)) {
+		WT_RET(__wt_txn_begin(session, txn_cfg));
+		started_txn = 1;
+	}
+	F_SET(txn, WT_TXN_READONLY);
+
+	/* Save a copy of the transaction's snapshot. */
+	WT_ERR(__wt_calloc_one(session, &nsnap_new));
+	nsnap = nsnap_new;
+	WT_ERR(__wt_strndup(
+	    session, cval.str, cval.len, &nsnap->name));
+	nsnap->snap_min = txn->snap_min;
+	nsnap->snap_max = txn->snap_max;
+	if (txn->snapshot_count > 0) {
+		WT_ERR(__wt_calloc_def(
+		    session, txn->snapshot_count, &nsnap->snapshot));
+		memcpy(nsnap->snapshot, txn->snapshot,
+		    txn->snapshot_count * sizeof(*nsnap->snapshot));
+	}
+	nsnap->snapshot_count = txn->snapshot_count;
+
+	/* Update the list. */
+
+	/*
+	 * The semantic is that a new snapshot with the same name as an
+	 * existing snapshot will replace the old one.
+	 */
+	WT_ERR_NOTFOUND_OK(__nsnap_drop_one(session, &cval));
+
+	if (STAILQ_EMPTY(&txn_global->nsnaph))
+		txn_global->nsnap_oldest_id = nsnap_new->snap_min;
+	STAILQ_INSERT_TAIL(&txn_global->nsnaph, nsnap_new, q);
+	nsnap_new = NULL;
+
+err:	if (started_txn)
+		WT_TRET(__wt_txn_rollback(session, NULL));
+	else if (ret == 0)
+		F_SET(txn, WT_TXN_NAMED_SNAPSHOT);
```

Comment:
This block of code deserves some comment. The crux is:

If an application called ```WT_SESSION::snapshot``` and there wasn't a transaction running, we leave without a transaction running. If there was a transaction running, we leave with the transaction running (i.e: the application needs to call commit or rollback on that transaction.

We could make it that calling WT_SESSION::snapshot would end the current transaction - an application can always do a ```begin_transaction(snapshot=XX)```. I suspect that's not intuitive.